File: packages/js-sdk/src/regex/detectors/uk/ni-number.ts
// UK National Insurance Number detector (scrubadub-inspired pattern)

import { RegexDetector } from '../../base'
import { EntityType } from '../../entities'
import { registerRegion } from '../../registry'

class NiNumberDetector extends RegexDetector {
  entityType = EntityType.NI_NUMBER
  score = 0.9
  // Scrubadub pattern: negative lookaheads for invalid prefixes
  pattern = new RegExp(
    '\\b' +
      '(?!BG)(?!GB)(?!NK)(?!KN)(?!TN)(?!NT)(?!ZZ)' +
      '[A-CEGHJ-PR-TW-Z][A-CEGHJ-NPR-TW-Z]' +
      '\\s?\\d{2}\\s?\\d{2}\\s?\\d{2}\\s?[A-D]' +
      '\\b',
    'gi'
  )
}

registerRegion('uk', NiNumberDetector)

export { NiNumberDetector }

File: packages/python-sdk/src/blindfold/regex/detectors/uk/ni_number.py
"""UK National Insurance Number detector (scrubadub-inspired pattern)"""

import re

from ...base import RegexDetector
from ...entities import EntityType
from ...registry import register_region


@register_region("uk")
class NiNumberDetector(RegexDetector):
    entity_type = EntityType.NI_NUMBER
    score = 0.90
    # Scrubadub pattern: negative lookaheads for invalid prefixes
    pattern = re.compile(
        r"\b"
        r"(?!BG)(?!GB)(?!NK)(?!KN)(?!TN)(?!NT)(?!ZZ)"
        r"[A-CEGHJ-PR-TW-Z][A-CEGHJ-NPR-TW-Z]"
        r"\s?\d{2}\s?\d{2}\s?\d{2}\s?[A-D]"
        r"\b",
        re.IGNORECASE,
    )
